Feature/add cuda

corentin-dev opened the pull request:
This pull request is to add CUDA (or other AbstractVector) to PencilArrays. It is meant as a base for improvements.

It should create all the underlying structure with the good data types. I had some bugs with null pointers to CUDA arrays, that were solved with an ugly workaround (using a minimum resize of 1).

Indexing is still a problem for GPU, and it should be considered to use `permdims` in `copy_range!` for instance.

jipolanco commented:
Thanks a lot for this work! It would be great if we could make things work with CUDA. Unfortunately I can't test on GPUs right now...

For now I have just some concerns regarding type stability, but I think these are easy to fix.

I also agree on the necessity of avoiding indexing for GPU arrays. Maybe, to avoid depending on CUDA, functions like `copy_range!` could have a default implementation (e.g. for `AbstractVector`) that avoids indexing, and a more specific implementation for CPU `Vector`s that uses indexing.

jipolanco commented:
I hope you don't mind, but I made a couple of changes to make sure that things are type stable. I also added some tests.

One of the things that I changed is the way the array type is specified in the `Pencil` constructor. Instead of taking a `array_type` keyword argument, now `Pencil` accepts an optional positional argument at the beginning that can be `Array`, `CuArray`, etc... For instance, this should work as expected:

```julia
p = Pencil(CuArray, (8, 10), MPI.COMM_WORLD)
u = PencilArray{Float32}(undef, p)
```

This way I could get things to be type stable, and in my opinion it's also simpler to use (but this is subjective!).

I also changed the names of the `eltype_*` functions, which are no longer exported. (I didn't know about `.name.wrapper`, nice trick! Too bad that Julia doesn't export a function to do this). Some changes will be needed in your PencilFFTs PR, but I guess it's better if we finish this PR first.

corentin-dev commented:
> I hope you don't mind, but I made a couple of changes to make sure that things are type stable. I also added some tests.

Perfect!

> One of the things that I changed is the way the array type is specified in the `Pencil` constructor. Instead of taking a `array_type` keyword argument, now `Pencil` accepts an optional positional argument at the beginning that can be `Array`, `CuArray`, etc... For instance, this should work as expected:
> 
> ```julia
> p = Pencil(CuArray, (8, 10), MPI.COMM_WORLD)
> u = PencilArray{Float32}(undef, p)
> ```
> This way I could get things to be type stable, and in my opinion it's also simpler to use and more idiomatic (but this is subjective!).

Yes, it makes sense. I did not want to change too much the code, but it is better this way.

> I also changed the names of the `eltype_*` functions, which are no longer exported. (I didn't know about `.name.wrapper`, nice trick! Too bad that Julia doesn't export a function to do this). Some changes will be needed in your PencilFFTs PR, but I guess it's better if we finish this PR first.

Ok, I'll have to change some things in PencilFFTs since I used it for some change for CUDA too. I'll let you know.
